torch/csrc/onnx/export.cpp

```diff
@@ -256,18 +256,48 @@ static void addAttribute(onnx::NodeProto * n_p, jit::Node * n, jit::Symbol name)
       break;
     case AttributeKind::g: {
       auto g = attr->mutable_g();
-      encodeGraph(g, n->g(name), {});
+      encodeGraph(g, n->g(name), {}, {}, {});
     } break;
     case AttributeKind::gs:
       for(auto & v : n->gs(name)) {
         auto g = attr->add_graphs();
-        encodeGraph(g, v, {});
+        encodeGraph(g, v, {}, {}, {});
       }
       break;
   }
 }
 
-static void encodeGraph(onnx::GraphProto * p_g, std::shared_ptr<Graph> & g, const std::vector<at::Tensor> & initializers) {
+
+static void encodeGraph(onnx::GraphProto * p_g, std::shared_ptr<Graph> & g,
+  const std::vector<std::string> & input_names,
+  const std::vector<std::string> & output_names,
+  const std::vector<at::Tensor> & initializers) {
+
+  std::unordered_map<Node *, std::string> node_names;
+  auto initNames = [&](const node_list & nodes, const std::vector<std::string> & names) {
+    JIT_ASSERT(names.size() <= nodes.size());
+    size_t idx = nodes.size() - names.size();
+    for (auto & name : names) {
+      auto node = nodes.at(idx++);
+      auto it = node_names.find(node);
+      // rarely, output nodes may be repeated
+      // if this happens, we just fail since using output_names is ambiguous in this case
+      if(it != node_names.end()) {
+        jit::barf("output '%s' and '%s' are the same node, so they cannot have the same output name",
+        name.c_str(), it->second.c_str());
+      }
+      node_names[node] = name;
+    }
+  };
+  initNames(g->inputs(), input_names);
+  initNames(g->outputs(), output_names);
```

Comment:
Can you move this init to a separate function that returns `node_names`? There's no reason it has to stay in the body of this huge thing